omniledger/service/service.go

```diff
@@ -809,14 +811,22 @@ clientTransactions:
 			}
 			states = append(states, scs...)
 			cin = cout
+
+		}
+		// timeout is ONLY used when the leader calls createStateChanges as
+		// part of planning which ClientTransactions fit into one block.
+		if timeout != noTimeout {
+			if time.Now().After(deadline) {
+				return
+			}
 		}
 		cdbTemp = cdbI.c
 		ctsOK = append(ctsOK, ct)
 	}
 
 	// Store the result in the cache before returning.
 	merkleRoot = cdbTemp.GetRoot()
-	s.stateChangeCache.update(scID, cts.Hash(), merkleRoot, ctsOK, states)
+	s.stateChangeCache.update(scID, cts.Hash(), merkleRoot, ctsOK, ctsBad, states)
```

Comment:
I think something interesting is going on here when we return early. If the function times out, then `cts != ctsOK` and we store a state change cache keyed on `cts.Hash()`. Then when we do `createNewBlock`, this function gets called again but on `ctsOK` (or `txOut` in your code), so it'll look for a key on named `ctsOK.Hash()`. OL won't be able to use the data cached from the first call to `createNewBlock`. There won't be a wrong behaviour, there will simply be a cache miss if a timeout happens. We probably don't need to change any of the code, but put a comment to explain that this may happen.

Reply:
Actually, the cache will be invalid anyway because `createNewBlock` will re-order the transactions. I have a PR that removes the re-ordering - https://github.com/dedis/cothority/pull/1389, will rebase to this PR when it's in master.